pkg/keyspace/tso_keyspace_group.go

```diff
@@ -156,6 +170,38 @@ func (m *GroupManager) Close() {
 	m.wg.Wait()
 }
 
+// patrolKeyspaceAssignment is used to patrol all keyspaces and assign them to the keyspace groups.
+func (m *GroupManager) patrolKeyspaceAssignment() error {
+	m.Lock()
+	defer m.Unlock()
+	if m.patrolKeyspaceAssignmentOnce {
+		return nil
+	}
+	keyspaces, err := m.store.LoadRangeKeyspace(utils.DefaultKeyspaceID, 0)
+	if err != nil {
+		return err
+	}
+	config, err := m.getKeyspaceConfigByKindLocked(endpoint.Basic)
+	if err != nil {
+		return err
+	}
+	for _, ks := range keyspaces {
+		if ks == nil {
+			continue
+		}
+		groupID, err := strconv.ParseUint(config[TSOKeyspaceGroupIDKey], 10, 64)
+		if err != nil {
+			return err
+		}
+		err = m.updateKeyspaceForGroupLocked(endpoint.Basic, groupID, ks.GetId(), opAdd)
+		if err != nil {
+			return err
```

Comment:
SaveKeypsaceGroups transaction could fail and return here, which leaves none or some of keyspaces being assigned with keyspace groups.  Because patrolKeyspaceAssignmentOnce is still false, next split will trigger patrolKeyspaceAssignment again. For those keyspaces which have been assigned groups in the last round, because we don't persistent group assignment in keyspace meta and don't check if it's already assigned, we could pick a different keyspace group and assign this keyspace to it, which results in this keyspace being assigned to multiple groups.

Reply:
Actually every tso restart then becoming leader could cause a keyspace being assigned to multiple keyspace groups.